modules/clusterer/clusterer.c

```diff
@@ -21,1570 +21,1628 @@
-#include <stdio.h>
-#include <assert.h>
-#include <unistd.h>
-#include "../../sr_module.h"
 #include "../../str.h"
 #include "../../dprint.h"
-#include "../../usr_avp.h"
-#include "../../db/db.h"
-#include "../../socket_info.h"
-#include "../../resolve.h"
 #include "../../mem/mem.h"
 #include "../../mem/shm_mem.h"
-#include "../../rw_locking.h"
-#include "../../error.h"
-#include "../../ut.h"
-#include "../../mi/mi.h"
-#include "../../timer.h"
+#include "../../locking.h"
 #include "../../bin_interface.h"
+#include "../../timer.h"
 #include "../../forward.h"
-#include "clusterer.h"
-#include "api.h"
-
-#define DB_CAP DB_CAP_QUERY | DB_CAP_UPDATE
-
-#define check_val( _col, _val, _type, _not_null, _is_empty_str) \
-	do { \
-		if ((_val)->type!=_type) { \
-			LM_ERR("column %.*s has a bad type\n", _col.len, _col.s); \
-			goto error; \
-		} \
-		if (_not_null && (_val)->nul) { \
-			LM_ERR("column %.*s is null\n", _col.len, _col.s); \
-			goto error; \
-		} \
-		if (_is_empty_str && !VAL_STRING(_val)) { \
-			LM_ERR("column %.*s (str) is empty\n", _col.len, _col.s); \
-			goto error; \
-		} \
-	} while (0)
-
-/* lock */
-static rw_lock_t *ref_lock;
-
-/* time interval */
-static unsigned int prob_interval = 1;
 
+#include "api.h"
+#include "node_info.h"
+#include "clusterer.h"
 
 struct clusterer_binds clusterer_api;
 
-/* Database variables */
+struct mod_registration *clusterer_reg_modules;
+enum sip_protos clusterer_proto = PROTO_BIN;
 
-/* DB handler */
-static db_con_t *db_hdl;
-/* DB functions */
-static db_func_t dr_dbf;
+extern int ping_interval;
+extern int node_timeout;
+extern int ping_timeout;
 
-/* DB URL */
-str clusterer_db_url;
-/* DB TABLE */
-str db_table = str_init("clusterer");
+static str tmp_buff;
 
-/* db_table columns */
+static int set_link(clusterer_link_state new_ls, node_info_t *node_a, node_info_t *node_b);
+static int set_link_for_current(clusterer_link_state new_ls, node_info_t *node);
+static void call_cbs_event(cluster_info_t *clusters, int *clusters_to_call, int no_clusters);
 
-/* PK column */
-str id_col = str_init("id");
-str cluster_id_col = str_init("cluster_id");
-str machine_id_col = str_init("machine_id");
-str url_col = str_init("url");
-str state_col = str_init("state");
-
-str last_attempt_col = str_init("last_attempt");
-str duration_col = str_init("duration");
-str failed_attempts_col = str_init("failed_attempts");
-str no_tries_col = str_init("no_tries");
+inline void heartbeats_timer(void)
+{
+	struct timeval now;
+	utime_t last_ping_int;
+	utime_t ping_reply_int;
+	cluster_info_t *clusters_it;
+	node_info_t *node;
+	static str module_name = str_init("clusterer");
+	str send_buffer;
+	int init_buffer;
+	int check_call_cbs_event[MAX_NO_CLUSTERS] = {0};
+	int no_clusters = 0;
 
-str description_col = str_init("description");
-static db_key_t *clusterer_cluster_id_key;
-static db_val_t *clusterer_cluster_id_value;
+	gettimeofday(&now, NULL);
 
-static db_op_t op_eq = OP_EQ;
+	lock_get(ref_lock);
```

Comment:
I don't see where this lock is released - if the function is supposed to have the lock taken after it ends - that should be documented in the function's header.